File: examples/undocumented/libshogun/evaluation_cross_validation_mkl_weight_storage.cpp
/*
 * This software is distributed under BSD 3-clause license (see LICENSE file).
 *
 * Authors: Giovanni De Toni, Heiko Strathmann, Thoralf Klein
 */

#include <shogun/base/range.h>
#include <shogun/classifier/mkl/MKLClassification.h>
#include <shogun/classifier/svm/LibSVM.h>
#include <shogun/evaluation/ContingencyTableEvaluation.h>
#include <shogun/evaluation/CrossValidation.h>
#include <shogun/evaluation/StratifiedCrossValidationSplitting.h>
#include <shogun/features/DenseFeatures.h>
#include <shogun/kernel/CombinedKernel.h>
#include <shogun/kernel/GaussianKernel.h>
#include <shogun/labels/BinaryLabels.h>
#include <shogun/lib/parameter_observers/ParameterObserverCV.h>
#include <shogun/mathematics/Statistics.h>

using namespace shogun;

void gen_rand_data(SGVector<float64_t> lab, SGMatrix<float64_t> feat,
		float64_t dist)
{
	index_t dims=feat.num_rows;
	index_t num=lab.vlen;

	for (int32_t i=0; i<num; i++)
	{
		if (i<num/2)
		{
			lab[i]=-1.0;

			for (int32_t j=0; j<dims; j++)
				feat(j, i)=Math::random(0.0, 1.0)+dist;
		}
		else
		{
			lab[i]=1.0;

			for (int32_t j=0; j<dims; j++)
				feat(j, i)=Math::random(0.0, 1.0)-dist;
		}
	}
	lab.display_vector("lab");
	feat.display_matrix("feat");
}

SGMatrix<float64_t> calculate_weights(
    ParameterObserverCV& obs, int32_t folds, int32_t run, int32_t len)
{
	int32_t column = 0;
	SGMatrix<float64_t> weights(len, folds * run);
	for (auto o : range(obs.get_num_observations()))
	{
		auto obs_storage = obs.get_observation(o);
		for (auto i : range(obs_storage->get<index_t>("num_folds")))
		{
			auto fold = obs_storage->get("folds", i);
			auto machine =
			    fold->get("trained_machine")->as<MKLClassification>();
			auto k = machine->get_kernel()->as<CombinedKernel>();
			auto w = k->get_subkernel_weights();

			/* Copy the weights inside the matrix */
			/* Each of the columns will represent a set of weights */
			for (auto j = 0; j < w.size(); j++)
			{
				weights.set_element(w[j], j, column);
			}

			column++;
		}
	}
	return weights;
}

void test_mkl_cross_validation()
{
	/* generate random data */
	index_t num=10;
	index_t dims=2;
	float64_t dist=0.5;
	SGVector<float64_t> lab(num);
	SGMatrix<float64_t> feat(dims, num);
	gen_rand_data(lab, feat, dist);

	/*create train labels */
	auto labels=std::make_shared<BinaryLabels>(lab);

	/* create train features */
	auto features=std::make_shared<DenseFeatures<float64_t>>(feat);

	/* create combined features */
	auto comb_features=std::make_shared<CombinedFeatures>();
	comb_features->append_feature_obj(features);
	comb_features->append_feature_obj(features);
	comb_features->append_feature_obj(features);

	/* create multiple gaussian kernels */
	auto kernel=std::make_shared<CombinedKernel>();
	kernel->append_kernel(std::make_shared<GaussianKernel>(10, 0.1));
	kernel->append_kernel(std::make_shared<GaussianKernel>(10, 1));
	kernel->append_kernel(std::make_shared<GaussianKernel>(10, 2));
	kernel->init(comb_features, comb_features);

	/* create mkl using libsvm, due to a mem-bug, interleaved is not possible */
	auto svm=std::make_shared<MKLClassification>(std::make_shared<LibSVM>());
	svm->set_interleaved_optimization_enabled(false);
	svm->set_kernel(kernel);

	/* create cross-validation instance */
	index_t num_folds=3;
	auto split=std::make_shared<StratifiedCrossValidationSplitting>(labels,
			num_folds);
	auto eval=std::make_shared<ContingencyTableEvaluation>(ACCURACY);
	auto cross=std::make_shared<CrossValidation>(svm, comb_features, labels, split, eval, false);

	/* add print output listener and mkl storage listener */
	auto mkl_obs = std::make_shared<ParameterObserverCV>(true);
	cross->subscribe(&mkl_obs);

	/* perform cross-validation, this will print loads of information */
	auto result=cross->evaluate();

	/* print mkl weights */
	auto weights = calculate_weights(*mkl_obs, num_folds, 1, 3);
	weights.display_matrix("mkl weights");

	/* print mean and variance of each kernel weight. These could for example
	 * been used to compute confidence intervals */
	Statistics::matrix_mean(weights, false).display_vector("mean per kernel");
	Statistics::matrix_variance(weights, false).display_vector("variance per kernel");
	Statistics::matrix_std_deviation(weights, false).display_vector("std-dev per kernel");

	/* Clear */
	mkl_obs->clear();

	/* again for two runs */
	cross->set_num_runs(2);
	result=cross->evaluate();

	/* print mkl weights */
	SGMatrix<float64_t> weights_2 = calculate_weights(*mkl_obs, num_folds, 2, 3);
	weights_2.display_matrix("mkl weights");

	/* print mean and variance of each kernel weight. These could for example
	 * been used to compute confidence intervals */
	Statistics::matrix_mean(weights_2, false)
	    .display_vector("mean per kernel");
	Statistics::matrix_variance(weights_2, false)
	    .display_vector("variance per kernel");
	Statistics::matrix_std_deviation(weights_2, false)
	    .display_vector("std-dev per kernel");

	/* clean up */
}

int main()
{
//	env()->io()->set_loglevel(MSG_DEBUG);

	test_mkl_cross_validation();

	return 0;
}

File: examples/undocumented/python/evaluation_cross_validation_mkl_weight_storage.py
#!/usr/bin/env python

# This software is distributed under BSD 3-clause license (see LICENSE file).
#
# Authors: Heiko Strathmann

from numpy.random import randn
from numpy import *

# generate some overlapping training vectors
num_vectors=5
vec_distance=1
traindat=concatenate((randn(2,num_vectors)-vec_distance,
	randn(2,num_vectors)+vec_distance), axis=1)
label_traindat=concatenate((-ones(num_vectors), ones(num_vectors)));

parameter_list = [[traindat,label_traindat]]

def evaluation_cross_validation_mkl_weight_storage(traindat=traindat, label_traindat=label_traindat):
    from shogun import BinaryLabels
    import shogun as sg
    import numpy as np

    # training data, combined features all on same data
    features=sg.create_features(traindat)
    comb_features=sg.create("CombinedFeatures")
    comb_features.add("feature_array", features)
    comb_features.add("feature_array", features)
    comb_features.add("feature_array", features)
    labels=BinaryLabels(label_traindat)

    # kernel, different Gaussians combined
    kernel=sg.create("CombinedKernel")
    kernel.add("kernel_array", sg.create("GaussianKernel", width=0.1))
    kernel.add("kernel_array", sg.create("GaussianKernel", width=1))
    kernel.add("kernel_array", sg.create("GaussianKernel", width=2))

    # create mkl using libsvm, due to a mem-bug, interleaved is not possible
    libsvm = sg.create("LibSVM")
    svm = sg.create("MKLClassification", svm=sg.as_svm(libsvm),
            interleaved_optimization=False, kernel=kernel)

    # splitting strategy for 5 fold cross-validation (for classification its better
    # to use "StratifiedCrossValidation", but the standard
    # "StratifiedCrossValidationSplitting" is also available
    splitting_strategy = sg.create(
        "StratifiedCrossValidationSplitting", labels=labels, num_subsets=5)

    # evaluation method
    evaluation_criterium=sg.create("ContingencyTableEvaluation", type="ACCURACY")

    # cross-validation instance
    cross_validation = sg.create(
        "CrossValidation", machine=svm, features=comb_features,
        labels=labels, splitting_strategy=splitting_strategy,
        evaluation_criterion=evaluation_criterium, num_runs=3)

    # append cross vlaidation output classes
    mkl_storage=sg.create("ParameterObserverCV")
    cross_validation.subscribe(mkl_storage)

    # perform cross-validation
    result=cross_validation.evaluate()

    # print mkl weights
    weights = []
    for obs_index in range(mkl_storage.get("num_observations")):
        obs = mkl_storage.get_observation(obs_index).get("cross_validation_run")
        for fold_index in range(obs.get("num_folds")):
            fold = obs.get("folds", fold_index)
            w = fold.get("trained_machine").get("kernel").get("combined_kernel_weight")
            weights.append(w)

    print("mkl weights during cross--validation")
    print(weights)

if __name__=='__main__':
	print('Evaluation CrossValidationClassification')
	evaluation_cross_validation_mkl_weight_storage(*parameter_list[0])
